### PR TITLE
Added login feature for chromebot

### LINKED ISSUE #139: To add a login feature
**I'm submitting a ...**  (check one with "x")
- [ ] bug report
- [s] feature request
- [ ] support request => Please do not submit support requests here, instead ask your query in out Gitter channel at https://gitter.im/fossasia/susi_server

**Current behavior:**
No login feature

**Expected behavior:**
To add a login feature

**Would you like to work on the issue?**
Yes
```

### PR DESCRIPTION
**Pull request by stealthanthrax**
Fixes #139 

#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [x] My branch is up-to-date with the Upstream `master` branch.
- [x] The unit tests pass locally with my changes 
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

- [x] All the functions created/modified in this PR contain relevant docstrings.

#### Short description of what this resolves:
Added the login feature

#### Changes proposed in this pull request:

screen shot 2017-11-06 at 6 23 03 pm

### CODE REVIEW

**Comment by stealthanthrax** [comment]
@zamhaq , i have used the authentication code form firefox-bot.

File: package.json [hunk 1]
```diff
@@ -2,13 +2,20 @@
   "name": "susi_chromebot",
   "version": "1.0.0",
   "description": "Chrome Extension of SUSI AI",
-  "dependencies": {},
+    
+  "dependencies": {
+      "moment": "^2.18.1",
+    "react-cookie": "^2.0.7",
```
Comment (zamhaq) [inline package.json]: why `react-cookie` is in dependencies? This project has no react.

File: src/settings.html [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+	<br>
+	<h3 class="hr-heading">Settings</h3>
+	<select id="theme" name="theme">
+		<option value="light">Light</option>
+		<option value="dark">Dark</option>
+	</select>
+	<br>
+	<button class="clear-history-button" id="clearMessageHistory">Clear Message History</button>
+	<script src="jquery/jquery-3.2.1.min.js"></script>
+	<script src="scripts/main.js"></script>
```
Comment (zamhaq) [inline src/settings.html]: where is this `scripts/main.js` file? There is no such file here.

File: src/options.js [hunk 3]
```diff
@@ -0,0 +1,509 @@
...
+		url: loginEndPoint,
+		dataType: "jsonp",
+		jsonpCallback: "p",
+		jsonp: "callback",
+		crossDomain: true,
+		success: function (response) {
+			if(response.accepted){
+				accessToken = response.access_token;
+				userEmail = email;
+				browser.storage.sync.set({
```
Comment (zamhaq) [inline src/options.js]: For chromebot accessing browser storage is done as `chrome.storage.sync.set`. Did you copied all the things? Please fix these things.

**Comment by stealthanthrax** [comment]
@zamhaq ,I am new to backend , so please pardon the mistakes.And i'll rectify the code as per your comments.

File: src/options.js [hunk 4]
```diff
@@ -0,0 +1,513 @@
...
+		}
+		else{
+			showLoggedInBlock(false);
+		}
+
+	});
+}
+
+
+function saveOptions(e) {
```
[the remark this answers is not in the document]
Comment (stealthanthrax) [reply src/options.js]: @zamhaq ,if i remove this function ,login form will not submit

File: src/options.js [hunk 5]
```diff
@@ -0,0 +1,513 @@
...
+var userMapObj={latitude:null,longitude:null,status:null,mapids:[]};
+
+themeSelect.addEventListener("change", saveOptions);
+loginForm.addEventListener("submit", login);
+changePasswordForm.addEventListener("submit", handleChangePassword);
+logoutButton.addEventListener("click", logout);
+clearMessageHistoryButton.addEventListener("click", clearMessageHistory);
+
+document.addEventListener("DOMContentLoaded", persistSettings);
+function createSusiMessageTableHistory(tableData,columns,columnsData,timeString,msgId_susi){
```
[the remark this answers is not in the document]
Comment (stealthanthrax) [reply src/options.js]: @zamhaq ,same is the case with this one.

File: src/options.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (zamhaq) [reply src/options.js]: @stealthanthrax It says that saveOptions() is defined but never used. Can you explain why removal of this function will cause such issues?

File: src/options.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (zamhaq) [reply src/options.js]: @stealthanthrax where are these function being called? In some other file using background pages? please check that.

**Comment by stealthanthrax** [comment]
@ParthS007 ,am working on fixing them and will squash the commits as soon as the errors as resolved

**Comment by stealthanthrax** [comment]
@zamhaq ,please have a look.

File: src/options.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (stealthanthrax) [reply src/options.js]: when I remove this function ,on clicking the login buttion , it refreshes the page only.

File: src/options.js [hunk 6]
```diff
@@ -0,0 +1,513 @@
...
+	chrome.storage.sync.remove("messagesHistory");
+	chrome.storage.sync.remove("userMapObj");
+	chrome.storage.sync.remove("loggedUser");
+	accessToken = "";
+	showLoggedInBlock(false);
+	window.location.reload();
+	history.go(0);
+	window.location.href=window.location.href;
+}
+function applyUserSettings(){
```
[the remark this answers is not in the document]
Comment (stealthanthrax) [reply src/options.js]: @zamhaq ,this line is being used on line 122.

File: src/options.js [hunk 7]
```diff
@@ -0,0 +1,513 @@
...
+	}
+	else{
+		noLoggedInBlock.style.display="block";
+		loggedInBlock.style.display="none";
+		document.getElementById("username").value = "";
+		document.getElementById("password").value = "";
+	}
+}
+
+function login(event){
```
[the remark this answers is not in the document]
Comment (stealthanthrax) [reply src/options.js]: This is used on multiple locations.

File: src/options.js [hunk 8]
```diff
@@ -0,0 +1,272 @@
+/* global $ */
+/*global messagesHistory, enableSync, applyTheme, userMapObj, htmlMsg*/
```
[the remark this answers is not in the document]
Comment (stealthanthrax) [reply src/options.js]: i have removed this

File: src/options.js [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+
+
+themeSelect.addEventListener("change", saveOptions);
+loginForm.addEventListener("submit", login);
+changePasswordForm.addEventListener("submit", handleChangePassword);
+logoutButton.addEventListener("click", logout);
+clearMessageHistoryButton.addEventListener("click", clearMessageHistory);
+
+document.addEventListener("DOMContentLoaded");
+function createSusiMessageTableHistory(tableData,columns,columnsData,timeString,msgIdSusi){
```
[the remark this answers is not in the document]
Comment (stealthanthrax) [reply src/options.js]: removed this

File: src/options.js [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+}
+
+
+
+
+
+
+
+
+function saveOptions(e) {
```
[the remark this answers is not in the document]
Comment (stealthanthrax) [reply src/options.js]: Can't remove this

File: src/options.js [hunk 11]
```diff
@@ -0,0 +1,272 @@
...
+		
+
+	}
+
+}
+
+
+
+
+function createSusiMessageRssHistory(answers,count,currentTimeString,msgId){
```
[the remark this answers is not in the document]
Comment (stealthanthrax) [reply src/options.js]: removed this

File: src/options.js [hunk 12]
```diff
@@ -0,0 +1,272 @@
...
+
+			// fire the api call to change settings value on server
+			$.ajax({
+
+				url: url,
+				dataType: "jsonp",
+				jsonpCallback: "q",
+				jsonp: "callback",
+				crossDomain: true,
+				success: function (response) {
```
[the remark this answers is not in the document]
Comment (stealthanthrax) [reply src/options.js]: removed it

File: src/options.js [hunk 13]
```diff
@@ -0,0 +1,272 @@
...
+		arrows:true,
+		dots:false,
+		infinite:true,
+		dotsClass:"slick-dots slick-dots-ul",
+
+	});
+
+}
+
+function urlDomain(data) {
```
[the remark this answers is not in the document]
Comment (stealthanthrax) [reply src/options.js]: done

File: src/options.js [hunk 14]
```diff
@@ -0,0 +1,272 @@
...
+	table =table.concat("</tr>");
+
+	for(i = 0 ; i < tableData.length ; i++){
+		table = table.concat("<tr>");
+
+		for(j= 0; j < columns.length ;  j++){
+			//check if such column value exists for that record
+			if(tableData[i][columns[j]]){
+				var cellData  = tableData[i][columns[j]];
+				cellData.replace(/https?:[/|.|\w]*/gi,function composeLink(link){
```
[the remark this answers is not in the document]
Comment (stealthanthrax) [reply src/options.js]: done

File: src/options.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (stealthanthrax) [reply src/options.js]: @zamhaq ,have a look at this

**Comment by stealthanthrax** [comment]
@zamhaq , all these functions are being used on subsequent lines.

**Comment by ms10398** [comment]
@stealthanthrax If you are haviing problem in fixing it Can I work on it and make a PR.

**Comment by stealthanthrax** [comment]
@ms10398 , i have tried to resolve the issues.You can please suggested improvements 👍

File: src/options.js [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+			if(response.accepted){
+				accessToken = response.accessTken;
+				userEmail = email;
+				chrome.storage.sync.set({
+					loggedUser:{
+						email:email,
+						accessToken: accessToken
+					}
+				});
+				time = response.valid_seconds;
```
[the remark this answers is not in the document]
Comment (ms10398) [reply src/options.js]: Correct this @stealthanthrax

**Comment by stealthanthrax** [comment]
@ms10398 @gabru-md ,done and squashed the commits .

**Comment by gabru-md** [comment]
@ms10398 please check if it is working and then review. I'm busy with exams so cannot review earlier than tomorrow.

File: src/options.js [hunk 16]
```diff
@@ -0,0 +1,146 @@
+/* global $ */
+/*global messagesHistory, enableSync,userMapObj, htmlMsg*/
```
[the remark this answers is not in the document]
Comment (ms10398) [reply src/options.js]: Please remove the messageHistory element if it is not used @stealthanthrax

File: src/options.js [hunk 17]
```diff
@@ -0,0 +1,146 @@
...
+logoutButton.addEventListener("click", function logout(e){
+    e.preventDefault();
+	chrome.storage.sync.remove("messagesHistory");
+	chrome.storage.sync.remove("loggedUser");
+});
+clearMessageHistoryButton.addEventListener("click", clearMessageHistory);
+
+document.addEventListener("DOMContentLoaded");
+
+function showLoggedInBlock(show){
```
[the remark this answers is not in the document]
Comment (ms10398) [reply src/options.js]: Please remove this also if it is not used @stealthanthrax

File: src/script.js [hunk 18]
```diff
@@ -12,11 +12,35 @@ var micmodal = document.getElementById("micmodal");
...
 var dark = false;
 var upCount = 0;
 var shouldSpeak = true;
 var storageItems= [];
 var storageArr = [];
+var inp = document.getElementById('inp');
+var box = document.getElementById('box');
+
+ function change (box,inp){
+    box.style.backgroundImage = 'url('+ inp.value + ')';
```
[the remark this answers is not in the document]
Comment (ms10398) [reply src/script.js]: Please keep the quotes even over all the code

File: src/script.js [hunk 19]
```diff
@@ -12,11 +12,35 @@ var micmodal = document.getElementById("micmodal");
 var setting = document.getElementById("setting");
 var clear = document.getElementById("clear");
 var settings = document.getElementById("settings");
+var scrollIconElement = document.getElementById("scrollIcon");
 var dark = false;
 var upCount = 0;
 var shouldSpeak = true;
 var storageItems= [];
 var storageArr = [];
+var inp = document.getElementById('inp');
```
[the remark this answers is not in the document]
Comment (ms10398) [reply src/script.js]: I think inp is never used so it should not be there

**Comment by ms10398** [comment]
Resolve all codacy errors @stealthanthrax

**Comment by stealthanthrax** [comment]
done @ms10398

**Comment by ms10398** [comment]
Resolve all Travis errors @stealthanthrax

**Comment by stealthanthrax** [comment]
@ms10398 ,the checks have passed.Please review.

**Comment by stealthanthrax** [comment]
@ms10398 I have done that.But i have made added a change of almost 400 lines and it is showing a commit of 100 lines ,so can i make a new PR for it? 
@ms10398 @gabru-md @zamhaq

**Comment by ms10398** [comment]
@stealthanthrax please update your fork with upstream master and please try if you can solve the issue in this PR.

**Comment by stealthanthrax** [comment]
@ms10398 , bro its not working ,login file is somehow deleted.

**Comment by ms10398** [comment]
Please try to solve all issues and make the PR ASAP as this issue has took very long time now and all fellow developers are waiting for it so they can work on other features related to this and make yourself familiar with git and GitHub workflow.

**Comment by stealthanthrax** [comment]
@ms10398 , even though my exams are **still going on** , i am trying be as active on the community as possible.
And I have made a new PR ,please have a look.

**Comment by stealthanthrax** [comment]
@zamhaq @ms10398 @gabru-md , I have added PR #188 , please have a look

**Pull request closed without merging**